include/nbl/asset/CCPUMeshPackerV2.h

```diff
@@ -0,0 +1,222 @@
+{
+
+// eventually I will rename it
+template <typename MDIStructType = DrawElementsIndirectCommand_t>
+class CCPUMeshPackerV2 final : public IMeshPacker<ICPUMeshBuffer, MDIStructType>
+{
+    using base_t = IMeshPacker<ICPUMeshBuffer, MDIStructType>;
+    using Triangle = typename base_t::Triangle;
+    using TriangleBatch = typename base_t::TriangleBatch;
+
+public:
+    using AllocationParams = MeshPackerBase::AllocationParamsCommon;
+
+    struct AttribAllocParams
+    {
+        size_t offset = INVALID_ADDRESS;
+        size_t size = 0ull;
+    };
+
+	struct ReservedAllocationMeshBuffers
+    {
+        uint32_t mdiAllocationOffset;
+        uint32_t mdiAllocationReservedSize;
+        uint32_t indexAllocationOffset;
+        uint32_t indexAllocationReservedSize;
+        AttribAllocParams attribAllocParams[SVertexInputParams::MAX_VERTEX_ATTRIB_COUNT];
+
+        inline bool isValid()
+        {
+            return this->mdiAllocationOffset != core::GeneralpurposeAddressAllocator<uint32_t>::invalid_address;
+        }
+    };
+
+    struct PackedMeshBuffer
+    {
+        core::smart_refctd_ptr<video::IGPUBuffer> MDIDataBuffer;
+        core::smart_refctd_ptr<video::IGPUBuffer> vertexBuffer;
+        core::smart_refctd_ptr<video::IGPUBuffer> indexBuffer;
+
+        inline bool isValid()
+        {
+            return this->MDIDataBuffer->getPointer() != nullptr;
+        }
+    };
+
+public:
+    CCPUMeshPackerV2(const AllocationParams& allocParams, uint16_t minTriangleCountPerMDIData = 256u, uint16_t maxTriangleCountPerMDIData = 1024u);
+
+	template <typename Iterator>
+	bool alloc(ReservedAllocationMeshBuffers* rambOut, const Iterator begin, const Iterator end);
+
+    void instantiateDataStorage();
+
+    template <typename Iterator>
+    bool commit(MeshPackerBase::PackedMeshBufferData* pmbdOut, ReservedAllocationMeshBuffers* rambIn, const Iterator begin, const Iterator end);
+
+    PackedMeshBuffer createGPUPackedMeshBuffer(video::IVideoDriver* driver);
+
+public:
+    core::smart_refctd_ptr<ICPUBuffer> m_vtxBuffer;
+    core::smart_refctd_ptr<ICPUBuffer> m_idxBuffer;
+    core::smart_refctd_ptr<ICPUBuffer> m_MDIBuffer;
+
+    const AllocationParams m_allocParams;
+
+};
+
+template <typename MDIStructType>
+CCPUMeshPackerV2<MDIStructType>::CCPUMeshPackerV2(const AllocationParams& allocParams, uint16_t minTriangleCountPerMDIData, uint16_t maxTriangleCountPerMDIData)
+    :IMeshPacker<ICPUMeshBuffer, MDIStructType>(minTriangleCountPerMDIData, maxTriangleCountPerMDIData),
+    m_allocParams(allocParams)
+{
+    initializeCommonAllocators(allocParams);
+}
+
+template <typename MDIStructType>
+template <typename Iterator>
+bool CCPUMeshPackerV2<MDIStructType>::alloc(ReservedAllocationMeshBuffers* rambOut, const Iterator begin, const Iterator end)
+{
+    size_t i = 0ull;
+    for (auto it = begin; it != end; it++)
+    {
+        ReservedAllocationMeshBuffers& ramb = *(rambOut + i);
+        const size_t vtxCnt = (*it)->calcVertexCount();
```

Comment:
this will need to be computed as a proper upper bound  because cretain vertices may need to be duplicated due to triangle batching

Reply:
same as for the case of packer V1, the upper bound is `(number of indices+1)/2`